src/Compression/CompressionCodecFSST.cpp

```diff
@@ -0,0 +1,144 @@
+#ifdef ENABLE_FSST
+
+#include <Compression/CompressionFactory.h>
+#include <Compression/CompressionInfo.h>
+#include <Compression/ICompressionCodec.h>
+#include <IO/VarInt.h>
+#include <base/types.h>
+
+#include <cstddef>
+#include <cstdint>
+#include <cstring>
+#include <stdexcept>
+
+#pragma clang diagnostic ignored "-Wcast-align"
+#pragma clang diagnostic ignored "-Wcast-qual"
+#pragma clang diagnostic ignored "-Wold-style-cast"
+#pragma clang diagnostic ignored "-Wimplicit-fallthrough"
+
+#include <fsst.h>
+
+namespace DB
+{
+
+class CompressionCodecFSST : public ICompressionCodec
+{
+public:
+    explicit CompressionCodecFSST() { setCodecDescription("FSST"); }
+
+    uint8_t getMethodByte() const override { return static_cast<uint8_t>(CompressionMethodByte::FSST); }
+
+    void updateHash(SipHash & hash) const override { getCodecDesc()->updateTreeHash(hash, /*ignore_aliases=*/true); }
+
+    static constexpr int out_size = 2281337;
+
+protected:
+    UInt32 doCompressData(const char * source, UInt32 source_size, char * dest) const override
+    {
+        std::vector<size_t> len_in;
+        std::vector<const unsigned char *> str_in;
+
+        splitDataByRows(reinterpret_cast<const unsigned char *>(source), str_in, len_in, source_size);
```

Comment:
If l. 39 would use `const char *`, we could remove all of the funny casting here and in the implementation of `splitDataByRows`.